Editorial review: Document CSS custom functions (#40703)
mdn/content: files/en-us/web/css/type/index.md

    - `<number>`
      - : Accepts any valid {{cssxref("&lt;number&gt;")}} value.
    - `<percentage>`
      - : Accepts any valid {{cssxref("&lt;percentage&gt;")}} value.
    - `<resolution>`
      - : Accepts any valid {{cssxref("&lt;resolution&gt;")}} value.
    - `<string>`
      - : Accepts any valid {{cssxref("&lt;string&gt;")}} value.
    - `<time>`
      - : Accepts any valid {{cssxref("&lt;time&gt;")}} value.
    - `<transform-function>`
      - : Accepts any valid {{cssxref("&lt;transform-function&gt;")}} value.
    - `<transform-list>`
      - : Accepts a list of valid {{cssxref("&lt;transform-function&gt;")}} values. It is equivalent to `"<transform-function>+"`, and may not be followed by a `+` or `#` token.
    - `<url>`
      - : Accepts any valid {{cssxref("url_value", "&lt;url&gt;")}} value.

- `*`
  - : The universal syntax.

You can combine `<syntax-type>` values using the following tokens:

- `+`
  - : A space-separated list of values is expected.
- `#`
  - : A comma-separated list of values is expected.

> In addition, the `|` token can be used to create loical OR conditions for the expected syntax, combining both `<ident>` and `<syntax-type>` values.

Reviewer: I’d suggest simplifying the description a little and also mentioning multiple values:

```suggestion
In addition, the `|` token can be used as a separator when specifying multiple values or combining `<ident>` and `<syntax-type>` values for the expected syntax.
```

### Return value

A data type definition.

## Description

The `type()` function is used when you need to define a data type. It can be considered as a subset of the overall value definition syntax used to define the set of valid values for every CSS property and function.

Most commonly, you'll see `type()` used to specify a single data type. The next example uses the {{cssxref("attr()")}} function to set the {{cssxref("background-color")}} property equal to the value of a user-defined `data-background` function. The required data type for the value has been specified as a {{cssxref("&lt;color>")}}.

```css
background-color: attr(data-background type(<color>), red);
```

You could also specify an exact keyword requirement (for example, `type(blue)`), but this would be overly limiting.

Specifying `type(*)` permits any valid CSS value. This is called the universal syntax, and it cannot be multiplied or combined with other syntax components.

### Specifying multiple permitted types

> You can use the `|` token to create OR logic and specify a range of permitted data types, keywords, or a combination of both. For example:

Reviewer: ```suggestion
You can use the `|` token as a separator when specifying a range of permitted data types, keywords, or a combination of both. For example:
```

```css
@function --color-choice(--color1 type(red | green), --color2 blue) {
  result: var(--color1);
  @media (width < 700px) {
    result: var(--color2);
  }
}
```

> [!NOTE]
> In the case of `@function` data types, you can omit the `type()` function and just include the value in cases where only one data type or keyword is specified. This is the case with the `blue` type definition in the previous custom function. This doesn't work with the `attr()` function.

### Specifying lists of types

The `+` and `#` tokens can be appended to a `<syntax-type>` to specify that you are expecting a space-separated list or a comma-separated list, respectively. For example:

- A `<color>+` parameter expects a space-separated list of `<color>` values, for example `red blue #a60000 rgb(234 45 100)`.
- A `<length>#` parameter expects a comma-separated lists of `<length>` values, for example `30px, 1em, 15vw`.

> You can combine OR logic with these tokens. For example, `<color># | <integer>#` would expect a comma-separated list of `<color>` values or a comma-separated list of `<integer>` values.

Reviewer: ```suggestion
You can combine multiple tokens using `|` as a separator. For example, `<color># | <integer>#` would expect a comma-separated list of `<color>` values or a comma-separated list of `<integer>` values.
```
